### PR TITLE
Fixed comparing compatible types for raw types

### PR DESCRIPTION
**Pull request by mlangkabel**
Until now we had an issue solving a method that requires a generic type as argument but is provided a raw type (see issue #68).
The implementation on this branch works fine but we lose the information that the type provided as argument is actually a raw type. Therefore I'm not sure if we should use this implementation or find a different solution. 
I think that the "different solution" would involve implementing edge cases in every method that can fail because of raw types (e.g. `ReferenceTypeUsage.compareConsideringTypeParameters(...)`, `GenericTypeInferenceLogic.consider(...)`, ...). 

So I don't know if this would be worth it. So I would propose to use the method implemented here and store the fact wether or not a ReferenceTypeUsage uses a raw type as a boolean field. 

@ftomassetti: What do you think?

### CODE REVIEW

**Comment by mlangkabel** [comment]
It seems like the tests are failing on travis while they are running fine on my machine. Is there a way to view the report the travis server generates?

**Comment by matozoid** [comment]
Just click "Details" in the box above.

**Comment by mlangkabel** [comment]
Thanks @matozoid. I did that but it just shows me the Travis job log. There it says:
```
There were failing tests. See the report at: file:///home/travis/build/ftomassetti/java-symbol-solver/java-symbol-solver-core/build/reports/tests/index.html
```
Do you know where I can find that file?

**Comment by matozoid** [comment]
Unfortunately no.

**Comment by matozoid** [comment]
I get an error when compiling this branch though:

Failed tests:   parseComments(me.tomassetti.symbolsolver.resolution.AnalyseJavaParserTest): Line 15 of src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt is different from what is expected expected:<... ==> java.util.Deque[]> but was:<... ==> java.util.Deque[<java.lang.Object>]>

**Comment by mlangkabel** [comment]
Ok, thank you again. I'll take a look at it when I'm back at work tomorrow.

**Comment by ftomassetti** [comment]
Maybe for now we could add `--info` and see if we get useful output:

```
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

Otherwise we could make the script write on the screen the content of the build directory which includes the detailed log.

**Comment by mlangkabel** [comment]
Ok, so the tests are running fine on Windows now. The only thing that keeps me from merging this pull request is that I'd like your feedback on storing the implicit type arguments of raw type as usage of the java.lang.Object type. Does this make sense?

File: java-symbol-solver-core/src/main/java/me/tomassetti/symbolsolver/model/typesystem/ReferenceTypeUsageImpl.java [hunk 1]
```diff
@@ -34,6 +35,12 @@ public ReferenceTypeUsageImpl(TypeDeclaration typeDeclaration, TypeSolver typeSo
 
     public ReferenceTypeUsageImpl(TypeDeclaration typeDeclaration, List<TypeUsage> typeParameters, TypeSolver typeSolver) {
         super(typeDeclaration, typeParameters, typeSolver);
+        if (!typeDeclaration.getTypeParameters().isEmpty() && typeParameters.isEmpty()) {
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/me/tomassetti/symbolsolver/model/typesystem/ReferenceTypeUsageImpl.java]: This super.typeParameters surprises me. Maybe you could instead have a static helper method which process the typeParameters and return them. You could call in this method in the super call:

```
super(typeDeclaration, myTypeParametersProcessor(typeDeclaration, typeParameters), typeSolver)
```

Comment (mlangkabel) [inline java-symbol-solver-core/src/main/java/me/tomassetti/symbolsolver/model/typesystem/ReferenceTypeUsageImpl.java]: Oh, that's a good idea! :)

File: java-symbol-solver-core/src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt [hunk 2]
```diff
@@ -12,7 +12,7 @@
   Line 51) CommentsCollection comments = new CommentsCollection() ==> com.github.javaparser.ast.comments.CommentsCollection
   Line 51) new CommentsCollection() ==> com.github.javaparser.ast.comments.CommentsCollection
   Line 52) int r ==> int
-  Line 54) Deque prevTwoChars = new LinkedList<Character>(Arrays.asList('z', 'z')) ==> java.util.Deque
+  Line 54) Deque prevTwoChars = new LinkedList<Character>(Arrays.asList('z', 'z')) ==> java.util.Deque<java.lang.Object>
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt]: Are these two equivalent? I think they are not because you can do this:

```
Box rawBox = new Box();           // rawBox is a raw type of Box<T>
Box<Integer> intBox = rawBox;     // warning: unchecked conversion
```

But you cannot do this:

```
Box<Object> rawBox = new Box<Object>();
Box<Integer> intBox = rawBox;
```

Comment (ftomassetti) [inline java-symbol-solver-core/src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt]: Probably not, we should add it. Unless we can consider that when there are no parameters is implictly a raw reference

Comment (mlangkabel) [inline java-symbol-solver-core/src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt]: The more I think about it, the more it feels like having a ReferenceTypeUsage that has no type arguments but has a declaration that has type parameters is implicitly a raw type. We could store it like that (and therefore revert the changes made on this branch). 
As a consequence we'd have to consider this case everytime we test if types are compatible.

Comment (mlangkabel) [inline java-symbol-solver-core/src/test/resources/javaparser_expected_output/com_github_javaparser_ast_comments_CommentsParser.txt]: I'll give it a try and close this pull request.
